frappe/contacts/doctype/address_template/address_template.py

```diff
@@ -4,52 +4,36 @@
 import frappe
 from frappe import _
 from frappe.model.document import Document
-from frappe.utils import cint
 from frappe.utils.jinja import validate_template
 
 
 class AddressTemplate(Document):
 	def validate(self):
+		validate_template(self.template)
+
 		if not self.template:
 			self.template = get_default_address_template()
 
-		self.defaults = frappe.db.get_values(
-			"Address Template", {"is_default": 1, "name": ("!=", self.name)}
-		)
-		if not self.is_default:
-			if not self.defaults:
-				self.is_default = 1
-				if cint(frappe.db.get_single_value("System Settings", "setup_complete")):
-					frappe.msgprint(_("Setting this Address Template as default as there is no other default"))
-
-		validate_template(self.template)
+		if not self.is_default and not self._get_previous_default():
+			self.is_default = 1
+			if frappe.db.get_single_value("System Settings", "setup_complete"):
+				frappe.msgprint(_("Setting this Address Template as default as there is no other default"))
 
 	def on_update(self):
-		if self.is_default and self.defaults:
-			for d in self.defaults:
-				frappe.db.set_value("Address Template", d[0], "is_default", 0)
+		if self.is_default and (previous_default := self._get_previous_default()):
```

Comment:
can this be clubbed with the condition in `validate`? maybe in a separate method?

Reply:
I tried for a bit, but didn't find a way to improve this. (Under the premise that `validate` should only validate and `on_update` can change other records.)

Reply:
we dont really need 2 db calls here ...we're basically doing get and then set.

better way would be to just set by directly passing the filters `{"is_default": 1, "name": ("!=", self.name)}` to set_value

`_get_previous_default` can then be removed.